### PR TITLE
Use git command instead of exec.Cmd in blame

### PR DESCRIPTION
**Pull request by lunny**
extract from #18147

### CODE REVIEW

File: modules/git/blame.go [hunk 1]
```diff
@@ -99,51 +95,39 @@ func (r *BlameReader) NextPart() (*BlamePart, error) {
...
-	// Here we use the provided context - this should be tied to the request performing the blame so that it does not hang around.
-	ctx, cancel, finished := process.GetManager().AddContext(ctx, fmt.Sprintf("GetBlame [repo_path: %s]", dir))
-
-	cmd := exec.CommandContext(ctx, command[0], command[1:]...)
-	cmd.Dir = dir
-	cmd.Stderr = os.Stderr
-	process.SetSysProcAttribute(cmd)
-
-	stdout, err := cmd.StdoutPipe()
+	cmd := NewCommandContextNoGlobals(ctx, "blame", CmdArg(commitID), "--porcelain")
```
Comment (wxiaoguang) [inline modules/git/blame.go]: There is no guarantee that `commitID` is always safe, so you need `CmdArgCheck` here, or use `AddDynamicArgument`

Comment (lunny) [reply modules/git/blame.go]: done.

File: modules/git/blame.go [hunk 2]
```diff
@@ -99,51 +95,39 @@ func (r *BlameReader) NextPart() (*BlamePart, error) {
...
-	// Here we use the provided context - this should be tied to the request performing the blame so that it does not hang around.
-	ctx, cancel, finished := process.GetManager().AddContext(ctx, fmt.Sprintf("GetBlame [repo_path: %s]", dir))
-
-	cmd := exec.CommandContext(ctx, command[0], command[1:]...)
-	cmd.Dir = dir
-	cmd.Stderr = os.Stderr
-	process.SetSysProcAttribute(cmd)
-
-	stdout, err := cmd.StdoutPipe()
+	cmd := NewCommandContextNoGlobals(ctx, "blame", CmdArgCheck(commitID), "--porcelain")
```
Comment (delvh) [inline modules/git/blame.go]: I think 
```suggestion
	cmd := NewCommandContextNoGlobals(ctx, "blame", AddDynamicArguments(commitID), "--porcelain")
```
is a better fit, as we then don't panic on an invalid commit, and instead handle the error more gracefully.

Comment (lunny) [reply modules/git/blame.go]: done.

**Review by delvh — APPROVED** [review]

**Comment by zeripath** [comment]
So IIRC the reasoning behind using exec.Cmd was to avoid the default timeout. I think we still need to do that.

**Comment by wxiaoguang** [comment]
> So IIRC the reasoning behind using exec.Cmd was to avoid the default timeout. I think we still need to do that.

The default timeout mechanism of git command seems not good.

And there is a `UseContextTimeout`,  see https://github.com/go-gitea/gitea/pull/20689

**Comment by lunny** [comment]
@zeripath @wxiaoguang done.

File: modules/git/blame.go [hunk 3]
```diff
@@ -99,51 +95,42 @@ func (r *BlameReader) NextPart() (*BlamePart, error) {
...
-
-	cmd := exec.CommandContext(ctx, command[0], command[1:]...)
-	cmd.Dir = dir
-	cmd.Stderr = os.Stderr
-	process.SetSysProcAttribute(cmd)
-
-	stdout, err := cmd.StdoutPipe()
+	cmd := NewCommandContextNoGlobals(ctx, "blame").
+		AddDynamicArguments(commitID).
+		AddArguments("--porcelain").
```
Comment (wxiaoguang) [inline modules/git/blame.go]:
```suggestion
	cmd := NewCommandContextNoGlobals(ctx, "blame", "--porcelain").
		AddDynamicArguments(commitID).
```

**Review by wolfogre — APPROVED** [review]

**Pull request merged**